Sources/TaskRunner.php

```diff
@@ -565,21 +561,35 @@ protected function performTask(array $task_details): bool
 		}
 
 		// Normally, the class should be specified using its fully qualified name.
-		if (class_exists($task_details['task_class']) && is_subclass_of($task_details['task_class'], 'SMF\\Tasks\\BackgroundTask')) {
+		if (
+			class_exists($task_details['task_class'])
+			&& is_subclass_of($task_details['task_class'], BackgroundTask::class)
+		) {
 			$details = empty($task_details['task_data']) ? [] : Utils::jsonDecode($task_details['task_data'], true);
 
 			$bgtask = new $task_details['task_class']($details);
 
+			if (!$bgtask->canExecute()) {
+				return true;
+			}
+
 			$success = $bgtask->execute();
 		}
 		// Just in case a mod or something specified a task without giving the namespace.
-		elseif (class_exists('SMF\\Tasks\\' . $task_details['task_class']) && is_subclass_of('SMF\\Tasks\\' . $task_details['task_class'], 'SMF\\Tasks\\BackgroundTask')) {
+		elseif (
+			class_exists('SMF\\Tasks\\' . $task_details['task_class'])
+			&& is_subclass_of('SMF\\Tasks\\' . $task_details['task_class'], BackgroundTask::class)
+		) {
 			$details = empty($task_details['task_data']) ? [] : Utils::jsonDecode($task_details['task_data'], true);
 
 			$task_class = 'SMF\\Tasks\\' . $task_details['task_class'];
 
 			$bgtask = new $task_class($details);
 
+			if (!$bgtask->canExecute()) {
+				return true;
+			}
+
 			$success = $bgtask->execute();
```

Comment:
This and the previous section are repeated.  Can we move this logic outside the if/elseif statements to reduce code?